### PR TITLE
snap: allow options for --strace, e.g. `snap run --strace="-tt"`

### PR DESCRIPTION
**Pull request by mvo5**
This allows passing extra options to snap run --strace.

Based on https://github.com/snapcore/snapd/pull/4473

### CODE REVIEW

File: cmd/snap/cmd_run.go [hunk 1]
```diff
@@ -423,7 +441,118 @@ func migrateXauthority(info *snap.Info) (string, error) {
...
+	gcmd := exec.Command(cmd[0], cmd[1:]...)
+	gcmd.Env = env
+	gcmd.Stdin = Stdin
+	gcmd.Stdout = Stdout
+	stderr, err := gcmd.StderrPipe()
+	if err != nil {
+		return err
+	}
+	filterDone := make(chan int)
+	var straceErrorOutput []string
```
Comment (bboozzoo) [inline cmd/snap/cmd_run.go]: `straceErrorOutput` does not appear to be used

Comment (mvo5) [reply cmd/snap/cmd_run.go]: Indeed, that is a merge artefact, I will remove it.

File: cmd/snap/cmd_run.go [hunk 2]
```diff
@@ -423,7 +441,116 @@ func migrateXauthority(info *snap.Info) (string, error) {
...
+				break
+			}
+			fmt.Fprint(Stderr, s)
+		}
+
+		// The last thing that snap-exec does is to
+		// execve() something inside the snap dir so
+		// we know that from that point on the output
+		// will be interessting to the user.
+		needle := fmt.Sprintf(`execve("%s`, dirs.SnapMountDir)
```
Comment (bboozzoo) [inline cmd/snap/cmd_run.go]:
```

10:20 | <mvo> | mborzecki: do you use /snap on your distro as prefix?
-- | -- | --
10:20 | <mvo> | mborzecki: still filtering 2: "[pid  6906] execve(\"/snap/hello-world/27/bin/echo\", [\"/snap/hello-world/27/bin/echo\"], 0xc8200b4a00 /* 73 vars */ <unfinished ...>\n" is the line I see - but *maybe* thats the issue
10:21 | <mvo> | mborzecki: i.e. that inside the snap env it is always /snap not dirs.SnapMountDir
10:21 | <zyga-ubuntu> | mvo: inside the env it's always /snap
10:21 | <zyga-ubuntu> | mvo: *except for classic
10:21 | <zyga-ubuntu> | on any distro
```

File: cmd/snap/cmd_run.go [hunk 3]
```diff
@@ -423,7 +441,116 @@ func migrateXauthority(info *snap.Info) (string, error) {
...
+	// ubuntu 14.04
+	var stracePath string
+	cand := filepath.Join(dirs.SnapMountDir, "strace-static", "current", "bin", "strace")
+	if osutil.FileExists(cand) {
+		stracePath = cand
+	}
+	if stracePath == "" {
+		stracePath, err = exec.LookPath("strace")
+		if err != nil {
+			return nil, fmt.Errorf("cannot find an installed strace, please try: `snap install strace-static`")
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap/cmd_run.go]: Not sure, we don't know if the $random-distro will have an strace that supports the syscalls we need to blacklist. So I think it is safer to suggest using the snap.

File: cmd/snap/cmd_run.go (hunk 2 above)
Comment (mvo5) [reply cmd/snap/cmd_run.go]: This is fixed now.

**Comment by zyga** [comment]
I'm marking this as blocked since the cause of the failures is very unclear for now.

File: cmd/snap/cmd_run.go [hunk 4]
```diff
@@ -276,7 +294,7 @@ func snapRunHook(snapName, snapRevision, hookName string) error {
 		return fmt.Errorf(i18n.G("cannot find hook %q in %q"), hookName, snapName)
 	}
 
-	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, nil)
+	return runSnapConfine(info, hook.SecurityTag(), snapName, "", hook.Name, runOptions{}, nil)
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: No stracing of hooks?

File: cmd/snap/cmd_run.go [hunk 5]
```diff
@@ -65,11 +68,12 @@ func init() {
...
-			"command": i18n.G("Alternative command to run"),
-			"hook":    i18n.G("Hook to run"),
-			"r":       i18n.G("Use a specific snap revision when running hook"),
-			"shell":   i18n.G("Run a shell instead of the command (useful for debugging)"),
-			"strace":  i18n.G("Run the command under strace (useful for debugging"),
+			"command":    i18n.G("Alternative command to run"),
+			"hook":       i18n.G("Hook to run"),
+			"r":          i18n.G("Use a specific snap revision when running hook"),
+			"shell":      i18n.G("Run a shell instead of the command (useful for debugging)"),
+			"strace":     i18n.G("Run the command under strace (useful for debugging). Extra strace options can be specified as well here."),
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: Nice, so this one option serves as as both the selector (don't use `strace` vs use `strace`) but also as a way to convey custom options. I think it warrants a comment for the casual reader.

File: cmd/snap/cmd_run.go [hunk 6]
```diff
@@ -51,11 +51,14 @@ var (
 
 type cmdRun struct {
 	Command  string `long:"command" hidden:"yes"`
-	Hook     string `long:"hook" hidden:"yes"`
+	Hookname string `long:"hook" hidden:"yes"`
```
Comment (zyga) [inline cmd/snap/cmd_run.go]: Nitpick: `HookName` or even move this to a separate PR
